Guard against thread safety issues related to assertions

pvollan opened the pull request:
#### 13958226016904e96c50fa14abcb35e708c699c4
<pre>
Guard against thread safety issues related to assertions
<a href="https://bugs.webkit.org/show_bug.cgi?id=268758">https://bugs.webkit.org/show_bug.cgi?id=268758</a>
<a href="https://rdar.apple.com/122321271">rdar://122321271</a>

Reviewed by Chris Dumez.

Guard against thread safety issues related to assertions, since they are usually acquired on a non-main thread.

* Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
(WebKit::ProcessAssertion::acquireSync):
(WebKit::ProcessAssertion::~ProcessAssertion):
* Source/WebKit/UIProcess/ProcessAssertion.h:

Canonical link: <a href="https://commits.webkit.org/274122@main">https://commits.webkit.org/274122@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0ef7f95ff17f5b1e5f53f905490d48c10c0dabf7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37975 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16881 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40230 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40520 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33783 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/19580 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14186 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32096 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38548 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14228 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/33246 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12423 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/12369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41793 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/34452 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34439 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38237 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12957 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/10598 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36426 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14523 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13380 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4928 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/13975 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/ProcessAssertion.h:
```diff
@@ -116,6 +116,7 @@ class ProcessAssertion : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<
 #if USE(EXTENSIONKIT)
+    static Lock s_capabilityLock;
     std::optional<AssertionCapability> m_capability;
     RetainPtr<_SEGrant> m_grant;
```
Should probably use `WTF_GUARDED_BY_LOCK` for this one and other members guarded by the lock to have the compiler verify we don't have threading bugs.

pvollan replied:
Ah, good point, addressed in latest patch.

cdumez commented on Source/WebKit/UIProcess/ProcessAssertion.h:
```diff
@@ -116,6 +116,7 @@ class ProcessAssertion : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<
     Function<void()> m_prepareForInvalidationHandler;
     Function<void()> m_invalidationHandler;
 #if USE(EXTENSIONKIT)
+    static Lock s_capabilityLock;
```
Why does the lock need to be static?

aproskuryakov replied:
And being static, is it even thread-safe?

pvollan replied:
> Why does the lock need to be static?

I made it static to guard agains potential threading issues in `[_SEExtensionProcess grantCapability: error:]`, if this is called simultaneously from multiple threads.

pvollan replied:
> And being static, is it even thread-safe?

That is a good point. I believe it is, even when being static. However, I will look closer into that.

pvollan replied:
Thanks for reviewing, all!

pvollan committed:
Guard against thread safety issues related to assertions
https://bugs.webkit.org/show_bug.cgi?id=268758
rdar://122321271

Reviewed by Chris Dumez.

Guard against thread safety issues related to assertions, since they are usually acquired on a non-main thread.

* Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
(WebKit::ProcessAssertion::acquireSync):
(WebKit::ProcessAssertion::~ProcessAssertion):
* Source/WebKit/UIProcess/ProcessAssertion.h:

Canonical link: https://commits.webkit.org/274122@main
